rocAL/source/readers/image/numpy_data_reader.cpp

```diff
@@ -0,0 +1,642 @@
+        _header_parsing_failed = true;
+        return;
+    }
+    token[n_read] = '\0';
+
+    // rocAL does not support numpy V2 headers
+    // https://numpy.org/neps/nep-0001-npy-format.html
+    int np_api_version = token[6];
+    if (np_api_version != 1) {
+        ERR("rocAL only supports reading npy files with NPY file format version 1");
+        _header_parsing_failed = true;
+        return;
+    }
+
+    // check if header is too short
+    std::string header = std::string(token.data());
+    if (header.find("NUMPY") == std::string::npos) {
+        ERR("File is not a numpy file");
+        _header_parsing_failed = true;
+        return;
+    }
+
+    // extract header length which can have up to 65535 bytes - NPYv1 format
+    uint16_t header_len = 0;
+    memcpy(&header_len, &token[8], 2);
+    if ((header_len + 10) % 16 != 0) {
+        ERR("Error extracting numpy header length");
+        _header_parsing_failed = true;
+        return;
+    }
+
+    token.resize(header_len + 1);
+    if (std::fseek(_current_file_ptr, offset, SEEK_SET)) {
+        ERR("Seek operation failed in " + file_path + ": " + std::strerror(errno));
+        _header_parsing_failed = true;
+        return;
+    }
+    n_read = std::fread(token.data(), 1, header_len, _current_file_ptr);
+    if (n_read != header_len) {
+        ERR("Can not read numpy header upto header_len");
+        _header_parsing_failed = true;
+        return;
+    }
+    token[header_len] = '\0';
+    header = std::string(token.data());
+    if (header.find('{') == std::string::npos) {
+        ERR("Header is corrupted");
+        _header_parsing_failed = true;
+        return;
+    }
+    offset += header_len;
+    if (std::fseek(_current_file_ptr, offset, SEEK_SET)) {
+        ERR("Seek operation failed in " + file_path + ": " + std::strerror(errno));
+        _header_parsing_failed = true;
+        return;
+    }
+
+    parse_header_data(parsed_header, header);
+    if (_header_parsing_failed) return;
+    parsed_header.data_offset = offset;
+}
+
+size_t NumpyDataReader::read_numpy_data(void* buf, size_t read_size, std::vector<unsigned>& strides_in_dims) {
+    if (!_current_file_ptr) {
+        ERR("Null file pointer");
+        return 0;
+    }
+
+    if (std::fseek(_current_file_ptr, _curr_file_header.data_offset, SEEK_SET)) {
+        ERR("Seek operation failed for " + _last_file_path + ": " + std::strerror(errno));
+        return 0;
+    }
+
+    auto shape = _curr_file_header.shape();
+
+    size_t actual_read_size = 0;
+
+    if (strides_in_dims[0] == _curr_file_header.size())
+        return std::fread((unsigned char*)buf, sizeof(unsigned char), _curr_file_header.numpy_data_nbytes(), _current_file_ptr);
+
+    if (_curr_file_header.type() == RocalTensorDataType::UINT8)
+        actual_read_size = parse_numpy_data<uint8_t>((uint8_t*)buf, strides_in_dims, shape);
+    else if (_curr_file_header.type() == RocalTensorDataType::UINT32)
```

Comment:
if you have a map table rocal_to_numpy_type(), you can replace the following if else statements with one line
parse_numpy_data<rocal_to_numpy_type(rocal_type)>

Reply:
Modified the function to read the data in bytes (sizeof(unsigned char) * dtype_size) and increment the data ptr accordingly